src/box/vy_scheduler.c

```diff
@@ -1116,6 +1117,140 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
 	return -1;
 }
 
+/**
+ * Table of symbols:
+ * R - Worst-case zero-result point lookup cost (in runs);
+ * W_u - number of unfiltered runs;
+ * W_f - number of filtered runs;
+ * t_r - Adopted run_size_ratio_coeff.
+ *
+ * Note:
+ * We should consider the fact that
+ * current range doesn't yet contains newly created run.
+ *
+ * @param lsm
+ * @param range
+ * @param new_run_stmts_count statements count in the newly created after
+ * dump/compaction run
+ * @param W total number of runs in the range
+ * after operation (dump/compaction)
+ * @param total_stmts_count total number of statements in the range
+ * after operation (dump/compaction)
+ * @param skip_n number of runs to skip during the range traverse
+ * as they are compacted into one new run.
+ * @return optimally calculated bloom_fpr
+ */
+double
+vy_task_find_range_bloom_fpr(
+	struct vy_lsm *lsm,
+	struct vy_range *range,
+	uint64_t new_run_stmts_count,
+	int W,
+	uint64_t total_stmts_count,
+	int skip_n)
+{
+	double R = W * lsm->opts.lookup_cost_coeff;
+
+	if (new_run_stmts_count == total_stmts_count) {
+		/** By definition of R. May also look at formulas ahead.
+		 *  R <= 1 in this case as soon as W = 1.
+		 *  It also means that skip_n == range->slice_count.
+		 */
+		return R;
+	}
+
+	struct vy_slice *last_slice =
+		rlist_last_entry(&range->slices, struct vy_slice, in_range);
+	uint64_t last_slice_stmts_count = last_slice->count.rows;
+
+	if (new_run_stmts_count > last_slice_stmts_count) {
+		/** The only case is when the size of dumped run
+		 *  is bigger than the size of the last run.
+		 *  Returning default bloom_fpr value,
+		 *  cause otherwise formula breaks.
+		 *  update_compaction_priority will understand that
+		 *  major compaction must be called and compaciton will create
+		 *  new run with optimal bloom_fpr
+		 */
+		return lsm->opts.bloom_fpr;
+	}
+
+	uint64_t scnd_last_slice_stmts_count = 0;
+
+	if (skip_n == range->slice_count - 1) {
+		/** In both dump and compaction cases it means
+		 *  we only have the new run except the last run.
+		 */
+		scnd_last_slice_stmts_count = new_run_stmts_count;
+	} else {
+		struct vy_slice *scnd_last_slice = rlist_prev_entry(
+			last_slice, in_range);
+		scnd_last_slice_stmts_count = scnd_last_slice->count.rows;
+	}
+
+	int W_u = 0;
+	int W_f = 0;
+	double t_r = 0;
+
+	if (R <= (total_stmts_count / last_slice_stmts_count)) {
+		W_u = 0;
+	} else {
+		t_r = (double)scnd_last_slice_stmts_count /
+		      last_slice_stmts_count;
```

Comment:
Why isn't `t_r >= 1` possible like in the case you described above?

Reply:
1.) If `scnd_last_slice_stmts_count` == `new_run_stmts_count`, this case is handled with returning default `bloom_fpr` value;
2.) If `scnd_last_slice_stmts_count` != `new_run_stmts_count`, situation when after the dump there exists a run **X**, which size is bigger than last run size, seems impossible: `vy_range_update_compaction_priority` must have called major compaction after **X** was dumped.

Reply:
Decided to add check that will make sure that none of existing runs are bigger than the last run.
Done.